ouroboros-consensus/src/Data/FingerTree/TopMeasured/Strict.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE BangPatterns           #-}
+{-# LANGUAGE ConstraintKinds        #-}
+{-# LANGUAGE DeriveGeneric          #-}
+{-# LANGUAGE DerivingStrategies     #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Data.FingerTree.TopMeasured.Strict (
+    StrictFingerTree
+    -- * Top-measuring
+  , InternalMeasured
+  , Measured (..)
+  , TopMeasured (..)
+  , measureInternal
+    -- * API
+  , empty
+  , fmap'
+  , fmap''
+  , fromList
+  , split
+  , (|>)
+  ) where
+
+import           Data.Foldable
+import           Data.Group
+import           GHC.Generics (Generic)
+
+import           NoThunks.Class (NoThunks (..), noThunksInValues)
+
+import           Data.FingerTree.Strict (Measured)
+import qualified Data.FingerTree.Strict as FT
+
+-- TODO(jdral): Should we force strictness anywhere? Or more generally, where
+-- should we guide evaluation?
+
+{-------------------------------------------------------------------------------
+  Strict finger trees with top-level measures
+-------------------------------------------------------------------------------}
+
+-- | A @StrictFingerTree@ with elements of type @a@, an internal measure
+-- of type @vi@, and a top-level measure of type @vt@.
+data StrictFingerTree vt vi a = StrictFingerTree  {
+    tm       :: vt
+  , elements :: !(FT.StrictFingerTree vi a)
+  }
+  deriving (Show, Eq, Ord, Generic)
+
+instance Foldable (StrictFingerTree vt vi) where
+  foldMap f = foldMap f . elements
+
+instance NoThunks a => NoThunks (StrictFingerTree vt vi a) where
+  showTypeOf _ = "Alt"
+  wNoThunks ctxt = noThunksInValues ctxt . toList
+
+instance (Semigroup vt, Measured vi a)
+      => Semigroup (StrictFingerTree vt vi a) where
+  StrictFingerTree tm1 xs1 <> StrictFingerTree tm2 xs2 =
+    StrictFingerTree (tm1 <> tm2) (xs1 <> xs2)
+
+instance (Monoid vt, Measured vi a) => Monoid (StrictFingerTree vt vi a) where
+  mempty = StrictFingerTree mempty mempty
+
+{-------------------------------------------------------------------------------
+  Measuring
+-------------------------------------------------------------------------------}
+
+-- | Internal measures are used by the internal @'StrictFingerTree'@.
+type InternalMeasured v a = Measured v a
+
+measureInternal :: Measured v a => a -> v
+measureInternal = FT.measure
+
+-- | All values of type @Alt@ are internal-measured.
```

Comment:
The `Alt` datatype is now called `StrictFingerTree`. We should ensure that all references to `Alt` are removed in comments (and other code)